source/common/event/file_event_impl.cc

```diff
@@ -117,11 +125,38 @@ void FileEventImpl::setEnabled(uint32_t events) {
     injected_activation_events_ = 0;
     activation_cb_->cancel();
   }
+  updateEvents(events);
+}
 
-  auto* base = event_get_base(&raw_event_);
-  event_del(&raw_event_);
-  assignEvents(events, base);
-  event_add(&raw_event_, nullptr);
+void FileEventImpl::unregisterEventIfEmulatedEdge(uint32_t event) {
+  ASSERT(event == FileReadyType::Write || event == FileReadyType::Read,
+         fmt::format("not allowed to unregister {}", event));
+  if (trigger_ == FileTriggerType::EmulatedEdge) {
+    auto new_event_mask = enabled_events_ & ~event;
+    if (event == FileReadyType::Write && injected_activation_events_ & FileReadyType::Read) {
+      // we cancel all the injected activation and merge the
+      // read event with the new registration mask
+      injected_activation_events_ = 0;
+      activation_cb_->cancel();
+      new_event_mask |= FileReadyType::Read;
+    }
+    updateEvents(new_event_mask);
+  }
+}
+
+void FileEventImpl::registerEventIfEmulatedEdge(uint32_t event) {
+  ASSERT(event == Event::FileReadyType::Write || event == Event::FileReadyType::Read);
+  if (trigger_ == FileTriggerType::EmulatedEdge) {
+    auto new_event_mask = enabled_events_ | event;
+    if (event == FileReadyType::Write && injected_activation_events_ & FileReadyType::Read) {
+      // we cancel all the injected activation and merge the
+      // read event with the new registration mask
+      injected_activation_events_ = 0;
```

Comment:
@antoniovicente would this cause a fairness issue? I am not sure about the implementation of libevent but  we cancel the `injected_activation_events_ ` and merge it with the registration mask. Would this push back the event on the dispatcher q?

Reply:
There's 2 issues here:  clearing the activated events may actually cause correctness issues.  I was mimicing libevent behavior when cleaning activate events when modifying the enabled mask, but I'm starting to think that's not a good idea.  Take https://github.com/envoyproxy/envoy/pull/13772 as a related issue; the loss of the requested read event after going through a read disable/re-enable cycle caused the connection to into a zombie state.  At least we're in the position now to change the activate event behavior as we need.  There's some current uses that I'ld like to tweak for performance, but I'm trying to sort out some implications on handling of close notifications and not having load tests that show performance implications of my proposed changes.

Reply:
I was thinking about it more while I was working on the `file_event_impl_tests`. For Level based events since we use `EVLOOP_ONCE` we should always use the legacy implementation of `FileEventImpl::activate` which does not set any `injected_activation_events_ `